src/Columns/MaskOperations.cpp

```diff
@@ -0,0 +1,320 @@
+          || extractMaskNumeric<inverted, UInt16>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, UInt32>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, UInt64>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, Int8>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, Int16>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, Int32>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, Int64>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, Float32>(mask, column, result, null_value, null_bytemap, nulls, mask_info)
+          || extractMaskNumeric<inverted, Float64>(mask, column, result, null_value, null_bytemap, nulls, mask_info)))
+        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot convert column {} to mask.", column->getName());
+
+    return mask_info;
+}
+
+template <bool inverted>
+MaskInfo extractMask(
+    const PaddedPODArray<UInt8> & mask,
+    const ColumnPtr & column,
+    PaddedPODArray<UInt8> & result,
+    UInt8 null_value)
+{
+    return extractMaskImpl<inverted>(mask, column, result, null_value, nullptr);
+}
+
+template <bool inverted>
+MaskInfo extractMaskInplace(
+    PaddedPODArray<UInt8> & mask,
+    const ColumnPtr & column,
+    UInt8 null_value)
+{
+    return extractMaskImpl<inverted>(mask, column, mask, null_value, nullptr);
+}
+
+template <bool inverted>
+MaskInfo extractMaskInplaceWithNulls(
+    PaddedPODArray<UInt8> & mask,
+    const ColumnPtr & column,
+    PaddedPODArray<UInt8> * nulls,
+    UInt8 null_value)
+{
+    return extractMaskImpl<inverted>(mask, column, mask, null_value, nullptr, nulls);
+}
+
+template MaskInfo extractMask<true>(const PaddedPODArray<UInt8> & mask, const ColumnPtr & column, PaddedPODArray<UInt8> & result, UInt8 null_value);
+template MaskInfo extractMask<false>(const PaddedPODArray<UInt8> & mask, const ColumnPtr & column, PaddedPODArray<UInt8> & result, UInt8 null_value);
+template MaskInfo extractMaskInplace<true>(PaddedPODArray<UInt8> & mask, const ColumnPtr & column, UInt8 null_value);
+template MaskInfo extractMaskInplace<false>(PaddedPODArray<UInt8> & mask, const ColumnPtr & column, UInt8 null_value);
+template MaskInfo extractMaskInplaceWithNulls<true>(PaddedPODArray<UInt8> & mask, const ColumnPtr & column, PaddedPODArray<UInt8> * nulls, UInt8 null_value);
+template MaskInfo extractMaskInplaceWithNulls<false>(PaddedPODArray<UInt8> & mask, const ColumnPtr & column, PaddedPODArray<UInt8> * nulls, UInt8 null_value);
+
+
+void inverseMask(PaddedPODArray<UInt8> & mask)
+{
+    std::transform(mask.begin(), mask.end(), mask.begin(), [](UInt8 val){ return !val; });
+}
```

Comment:
Here I would rather have ordinary for-loop (cause I not sure if `std::transform` can be vectorized by compiler).
Also, this function is used in a single place for now. I think we can move implementation there ...

Reply:
Good point, will fix